### PR TITLE
Fixes #244, #154, #196, #173, #235 and #239

### LINKED ISSUE #154: How to validate notEmpty bit value 0/1
I noticed that the NotEmpty rule uses PHP's empty(), which always evaluates the value of numeric 0 as false. Values of non-nullable 0 or 1 are common in bit-type fields where an on/off value is set. Any suggestions on how to validate this as non-empty?

### LINKED ISSUE #244: Int notempty should return true if 0
The validation of `v::int()->notEmpty(0)->validate()` returns false.

This is not intuitive.  0 is an integer, therefore it's not empty.  

Intuitively should work as follows:
```php
v::int()->notEmpty(0)->validate();    //true
v::int()->notEmpty('')->validate();    //false
v::int()->notEmpty('0')->validate(0);    //true
v::int()->notEmpty(null)->validate(0);    //false
```

### PR DESCRIPTION
**Pull request by alganet**
- Moves code from  __invoke on AbstractRule into NotEmpty.
  - Moves foreign type check to the chain.
  - Fixes tests using the chain instead of concrete instances.
  - Adds a parameter for strict type checking on NotEmpty. (no BC break)
  - Adds new tests for the BC behavior and new parameter.
  - Fixes AbstractRelated to allow empty referenced values.
  - Removes dead code found in #239 

Fixes #154, #196, #173 and #235.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -1475,14 +1475,19 @@ See also:
 
   * [v::noneOf()](#vnoneofv1-v2-v3)
 
-#### v::notEmpty()
+#### v::notEmpty($jugglesTypes = true)
```
Comment (henriquemoody) [inline README.md]: You must update the links too [links](https://github.com/Respect/Validation#vnotempty)

**Comment by henriquemoody** [comment]
@alganet, since you're doing this, cal you fix #235 too?

File: README.md [hunk 2]
```diff
@@ -1475,14 +1475,19 @@ See also:
...
 required. This function also takes whitespace into account, use `noWhitespace()`
 if no spaces or linebreaks and other whitespace anywhere in the input is desired.
 
+
 ```php
 v::string()->notEmpty()->validate(''); //false
+v::string()->notEmpty()->validate('0'); //false
+
+// Disable type juggling
+v::string()->notEmpty(false)->validate('0'); //true
```
Comment (henriquemoody) [inline README.md]: I was expecting `v::int()->notEmpty(false)->validate(0)` to return `true` too, but it didn't happen

File: library/Rules/NotEmpty.php [hunk 3]
```diff
@@ -3,12 +3,31 @@
...
+            return !empty($input);
         }
 
-        return !empty($input);
+        return '' !== $input;
+    }
+
+    public function __invoke($input)
+    {
+        return $this->validate($input);
```
Comment (henriquemoody) [inline library/Rules/NotEmpty.php]: What you did here and on `AbstractRule` was great! Very clever :clap:

File: library/Rules/NotEmpty.php [hunk 4]
```diff
@@ -3,12 +3,31 @@
...
+        $this->jugglesTypes = $jugglesTypes;
+    }
+
     public function validate($input)
     {
-        if (is_string($input)) {
-            $input = trim($input);
+        if (!is_string($input)) {
+            return !empty($input);
+        }
```
Comment (henriquemoody) [inline library/Rules/NotEmpty.php]: Just an idea for the first major version: when the input is an array we can use `array_filter()` to remove empty values.

File: README.md (hunk 2 above)
Comment (alganet) [inline README.md]: Do you think it should? That would make `notEmpty` a string-only rule.

**Comment by alganet** [comment]
I've added the fix for #235. It was actually a bug on `AbstractRelated` (the base class for Key, Attribute, Each and indirect validators).

Guys, I'm not sure about the parameter on `NotEmpty`. I've changed it to a simpler name. My implementation only disallows quoted zeroes like `'  0'` and `"0"` (with or without whitespace), which seems to be the cause of the trouble.

**Comment by henriquemoody** [comment]
Don't forget to update Validator docblock, too

File: README.md [hunk 5]
```diff
@@ -1476,13 +1476,19 @@ See also:
   * [v::noneOf()](#vnoneofv1-v2-v3)
 
 #### v::notEmpty()
+#### v::notEmpty($strict = false)
```
Comment (henriquemoody) [inline README.md]: Not a blocker, but I guess `v::notEmpty(boolean $strict)` is better.

File: README.md (hunk 2 above)
Comment (henriquemoody) [inline README.md]: My expectations in non-strict mode:

| Value             | NotEmpty |
| -------------     | ---------|
| `null`            | No       |
| `""`              | No       |
| `array()`         | Yes      |
| `"0"`             | Yes      |
| `0`               | Yes      |
| `"0.0"`           | Yes      |
| `0.0`             | Yes      |
| `true`            | Yes      |

Does it makes sense for you? Take a look on #205.

**Comment by henriquemoody** [comment]
#235 is not fixed yet, I think to fix that we have to overwrite all possible `validate()` methods...

File: library/Rules/AbstractRelated.php [hunk 6]
```diff
@@ -66,10 +66,12 @@ public function check($input)
...
     {
         $hasReference = $this->hasReference($input);
+
         if ($this->mandatory && !$hasReference) {
             return false;
         }
 
-        return $this->decision('validate', $hasReference, $input);
+        return $this->decision('validate', $hasReference, $input) ||
+        $this->getReferenceValue($input) === '';
```
Comment (malukenho) [inline library/Rules/AbstractRelated.php]: `empty()`

**Comment by henriquemoody** [comment]
I will close this pull request because `NotEmpty` has a new feature and new features must be send against to `master` branch.
Also I toke some part of code on this pull request and I sent #298.

About the NotEmpty, I was wondering about this, I'm in favor of creating a new rule, called `NotBlank` which should have more or less the same behavior of our current `NotEmpty` and the `NotEmpty` rule should use `empty()` but since it's a major BC it's better we do this in the 1.0 version, IMO.

Any way, thanks you very much for this pull request, #298 is already available on `0.7.5` and `0.8.1` versions and on `master` branch as well.

**Pull request closed without merging**